src/Microsoft.ML.Transforms/TensorflowTransform.cs

```diff
@@ -0,0 +1,665 @@
+                    return DataKind.R4;
+                }
+
+                if (type == TFDataType.Double)
+                {
+                    return DataKind.R8;
+                }
+
+                if (type == TFDataType.Int32)
+                {
+                    return DataKind.I4;
+                }
+
+                if (type == TFDataType.Int64)
+                {
+                    return DataKind.I8;
+                }
+
+                if (type == TFDataType.Bool)
+                {
+                    return DataKind.Bool;
+                }
+
+                if (type == TFDataType.String)
+                {
+                    return DataKind.TX;
+                }
+
+                throw new NotSupportedException("Tensorflow type not supported.");
+            }
+            private (ColumnType, TFDataType) BuildOuputMetaData(TFSession tfSession, int batchSize, string columnName)
+            {
+                var tfoutput = new TFOutput(tfSession.Graph[columnName]);
+                var shape = tfSession.Graph.GetTensorShape(tfoutput);
+
+                int[] dims = new int[shape.NumDimensions];
+                for (int k = 0; k < shape.NumDimensions; k++)
+                {
+                    dims[k] = (int)(shape[k] == -1 ? batchSize : shape[k]);
+                }
+                var kind = Tf2MlNetType(tfoutput.OutputType);
+               return (new VectorType(PrimitiveType.FromKind(kind), dims), tfoutput.OutputType);
+            }
+
+            private TFColInfo BuildTFInputMetaData(TFSession tfSession, int batchSize, string[] source)
+            {
+                var tfShapes = new TFShape[source.Length];
+                var tfTypes = new TFDataType[source.Length];
+                var colNames = new string[source.Length];
+                for (int j = 0; j < source.Length; j++)
+                {
+                    colNames[j] = source[j];
+                    var tfoutput = new TFOutput(tfSession.Graph[colNames[j]]);
+                    tfShapes[j] = tfSession.Graph.GetTensorShape(tfoutput);
+                    tfTypes[j] = tfoutput.OutputType;
+
+                    var l = new long[tfShapes[j].NumDimensions];
+                    for (int ishape = 0; ishape < tfShapes[j].NumDimensions; ishape++)
+                    {
+                        l[ishape] = tfShapes[j][ishape] == -1 ? batchSize : tfShapes[j][ishape];
+                    }
+                    tfShapes[j] = new TFShape(l);
+                }
+                return new TFColInfo(colNames, tfShapes, tfTypes);
+            }
+
+            protected override ColumnType GetColumnTypeCore(int iinfo)
+            {
+                Contracts.Assert(0 <= iinfo & iinfo < OutputCols.Length);
+
+                Contracts.Assert(OutputCols[iinfo] != null);
+                return OutputCols[iinfo];
+            }
+
+            private static string TestTypes(ColumnType[] types)
+            {
+                return null;
```

Comment:
>null [](start = 23, length = 4)

The IDV type system has key types, DateTime, unsigned integer types, etc. I think this should make sure that the types are one of the types allowed above in Tf2MlNetType().

Reply:
It also seems that the inputs must be vectors.

---
In reply to: [210646613](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210646613) [](ancestors = 210646613)

Reply:
I am not sure I was right about this assumption - does TF accept scalar inputs?

---
In reply to: [210657676](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210657676) [](ancestors = 210657676,210646613)